src/typechecker/FStar.TypeChecker.Util.fs

```diff
@@ -1011,26 +1011,96 @@ let check_trivial_precondition env c =
   
   ct, vc, Env.guard_of_guard_formula <| NonTrivial vc
 
-let maybe_coerce_bool_to_type env (e:term) (lc:lcomp) (t:term) : term * lcomp =
-    if env.is_pattern then e, lc else
+let coerce_with (env:Env.env) (e : term) (lc : lcomp) (f : lident) (ty : typ) : term * lcomp =
+    match Env.try_lookup_lid env f with
+    | Some _ ->
+        if Env.debug env (Options.Other "Coercions") then
+            BU.print1 "Coercing with %s!\n" (Ident.string_of_lid f);
+        let coertion = S.fvar (Ident.set_lid_range f e.pos) (Delta_constant_at_level 1) None in
+        let lc = bind e.pos env (Some e) lc (None, U.lcomp_of_comp <| S.mk_Total ty) in
+        let e = mk_Tm_app coertion [S.as_arg e] None e.pos in
+        e, lc
+    | None ->
+        Errors.log_issue e.pos (Errors.Warning_CoercionNotFound,
+                                (BU.format1 "Coercion %s was not found in the environment, not coercing."
+                                            (string_of_lid f)));
+        e, lc
+
+let maybe_coerce_lc env (e:term) (lc:lcomp) (t:term) : term * lcomp =
+    let is_t_term t =
+        let t = N.unfold_whnf env t in
+        match (SS.compress t).n with
+        | Tm_fvar fv -> S.fv_eq_lid fv C.term_lid
+        | _ -> false
+    in
+    let is_t_term_view t =
+        let t = N.unfold_whnf env t in
+        match (SS.compress t).n with
+        | Tm_fvar fv -> S.fv_eq_lid fv C.term_view_lid
+        | _ -> false
+    in
     let is_type t =
         let t = N.unfold_whnf env t in
         match (SS.compress t).n with
         | Tm_type _ -> true
         | _ -> false
     in
-    match (U.unrefine lc.res_typ).n with
-    | Tm_fvar fv
-        when S.fv_eq_lid fv C.bool_lid
-          && is_type t ->
-      let _ = Env.lookup_lid env C.b2t_lid in  //check that we have Prims.b2t in the context
-      let b2t = S.fvar (Ident.set_lid_range C.b2t_lid e.pos) (Delta_constant_at_level 1) None in
-      let lc = bind e.pos env (Some e) lc (None, U.lcomp_of_comp <| S.mk_Total (U.ktype0)) in
-      let e = mk_Tm_app b2t [S.as_arg e] None e.pos in
-      e, lc
+    let res_typ = U.unrefine lc.res_typ in
+    let head, args = U.head_and_args res_typ in
+    if Env.debug env (Options.Other "Coercions") then
+            BU.print3 "(%s) Trying to coerce from (%s) to (%s)\n"
+                    (Range.string_of_range e.pos)
+                    (Print.term_to_string res_typ)
+                    (Print.term_to_string t);
+
+    // checks if `t2 == erased t1`
+    let is_erased t1 t2 =
+        let head, args = U.head_and_args t2 in
+        match (U.un_uinst head).n, args with
+        | Tm_fvar fv, [(x, None)] ->
```

Comment:
Why do we insist on `aqual` as `None` for the `x` argument?

Reply:
Just being defensive.. there shouldn't be any other well-typed solution, but this is interleaved with typechecking so I don't want to rely on that.

Reply:
Not sure what do you mean, I just meant could we leave it unspecified, `[(x, _)]`, it is not clear to me that `aqual` can't take any other value at this point. But this is a minor comment, so `None` is also fine.